### PR TITLE
Add a64fx and fujitsu Compiler support

### PR DESCRIPTION
**Pull request by crtrott**
This adds support for A64FX (ARM v8.2 with SVE) and the Fujitsu Compiler in order to support Fugaku.

### CODE REVIEW

File: cmake/kokkos_compiler_id.cmake [hunk 1]
```diff
@@ -137,6 +137,17 @@ IF(KOKKOS_CXX_COMPILER_ID STREQUAL Cray OR KOKKOS_CLANG_IS_CRAY)
   ENDIF()
 ENDIF()
 
+IF(KOKKOS_CXX_COMPILER_ID STREQUAL Fujitsu)
+  # SET Cray's compiler version.
```
Comment (dalg24) [inline cmake/kokkos_compiler_id.cmake:141]: I assume you copied from above and forgot to update or remove the comment

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -18,7 +18,7 @@ TARGET_COMPILE_DEFINITIONS(kokkos_gtest PUBLIC GTEST_HAS_TR1_TUPLE=0 GTEST_HAS_P
 
 TARGET_INCLUDE_DIRECTORIES(kokkos_gtest PUBLIC ${GTEST_SOURCE_DIR})
 #Gtest minimally requires C++11
-IF(NOT (Kokkos_ENABLE_CUDA AND WIN32))
+IF((NOT (Kokkos_ENABLE_CUDA AND WIN32)) AND (NOT ("${KOKKOS_CXX_COMPILER_ID}" STREQUAL "Fujitsu")))
 TARGET_COMPILE_FEATURES(kokkos_gtest PUBLIC cxx_std_11)
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:22]: Would someone remind me why we need this in the first place?  Don't we set `CMAKE_CXX_STANDARD`?

File: cmake/kokkos_arch.cmake [hunk 3]
```diff
@@ -35,6 +35,7 @@ KOKKOS_ARCH_OPTION(ARMV80          HOST "ARMv8.0 Compatible CPU")
 KOKKOS_ARCH_OPTION(ARMV81          HOST "ARMv8.1 Compatible CPU")
 KOKKOS_ARCH_OPTION(ARMV8_THUNDERX  HOST "ARMv8 Cavium ThunderX CPU")
 KOKKOS_ARCH_OPTION(ARMV8_THUNDERX2 HOST "ARMv8 Cavium ThunderX2 CPU")
+KOKKOS_ARCH_OPTION(A64FX       HOST "ARMv8.2 with SVE Suport")
```
Comment (dalg24) [inline cmake/kokkos_arch.cmake:38]: Do not break the alignment.

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:22]: I thought someone found this was needed on some platform ...

File: cmake/kokkos_compiler_id.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_compiler_id.cmake:141]: yeah ...

File: cmake/kokkos_arch.cmake (hunk 3 above)
Comment (crtrott) [reply cmake/kokkos_arch.cmake:38]: bah that came through the renaming from ARMV8_SVE to A64FX based on feedback from @nmhamster . Gonna fix that,

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (masterleinad) [reply core/unit_test/CMakeLists.txt:22]: AFAIK, we don't compile `gtest` with the `Kokkos` flags and we don't require `CMAKE_CXX_STANDARD` directly. Hence, there were some issues in the nightlies at some point.

File: cmake/kokkos_arch.cmake (hunk 3 above)
Comment (crtrott) [reply cmake/kokkos_arch.cmake]: fixed

File: cmake/kokkos_compiler_id.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_compiler_id.cmake]: fixed

File: core/src/impl/Kokkos_TaskQueue_impl.hpp [hunk 4]
```diff
@@ -381,7 +381,7 @@ KOKKOS_FUNCTION void TaskQueue<ExecSpace, MemorySpace>::schedule_runnable(
 
   // If we don't have a dependency, or if pushing onto the wait queue of that
   // dependency failed (since the only time that queue should be locked is when
-  // the task is transitioning to complete??!?)
+  // the task is transitioning to complete?)
```
Comment (dalg24) [inline core/src/impl/Kokkos_TaskQueue_impl.hpp:384]: Drop this change

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (dalg24) [reply core/unit_test/CMakeLists.txt:22]: I would be surprised if we still need that line at all but this probably shouldn't be addressed here

File: cmake/kokkos_compiler_id.cmake [hunk 5]
```diff
@@ -137,6 +137,17 @@ IF(KOKKOS_CXX_COMPILER_ID STREQUAL Cray OR KOKKOS_CLANG_IS_CRAY)
...
+IF(KOKKOS_CXX_COMPILER_ID STREQUAL Fujitsu)
+  # SET Fujitsus compiler version which is not detected by CMake
+  EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} --version
+                  OUTPUT_VARIABLE INTERNAL_CXX_COMPILER_VERSION
+                  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+  STRING(REGEX MATCH "[0-9]+\\.[0-9]+\\.[0-9]+"
+         TEMP_CXX_COMPILER_VERSION ${INTERNAL_CXX_COMPILER_VERSION})
+  SET(KOKKOS_CXX_COMPILER_VERSION ${TEMP_CXX_COMPILER_VERSION} CACHE STRING INTERNAL FORCE)
+ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_compiler_id.cmake:149]: I don't think we actually need this block.  Why did you add it?

**Review by masterleinad — APPROVED** [review]
Looks OK to me (assuming that the flags are actually correct).

File: cmake/kokkos_compiler_id.cmake (hunk 5 above)
Comment (crtrott) [reply cmake/kokkos_compiler_id.cmake:149]: because otherwise KOKKOS_CXX_COMPILER_VERSION is empty and some code later fails.

File: core/src/impl/Kokkos_TaskQueue_impl.hpp (hunk 4 above)
Comment (dalg24) [inline core/src/impl/Kokkos_TaskQueue_impl.hpp:384]: From Christian
```
core/src/impl/Kokkos_TaskQueue_impl.hpp", line 384: warning: support for trigraphs is disabled
    // the task is transitioning to complete??!?)
```

Comment (dalg24) [reply core/src/impl/Kokkos_TaskQueue_impl.hpp:384]: From Christian
```
core/src/impl/Kokkos_TaskQueue_impl.hpp", line 384: warning: support for trigraphs is disabled
    // the task is transitioning to complete??!?)

**Review by dalg24 — APPROVED** [review]

File: algorithms/unit_tests/CMakeLists.txt [hunk 6]
```diff
@@ -24,7 +24,7 @@ KOKKOS_ADD_TEST_LIBRARY(
 # avoid deprecation warnings from MSVC
 TARGET_COMPILE_DEFINITIONS(kokkosalgorithms_gtest PUBLIC GTEST_HAS_TR1_TUPLE=0 GTEST_HAS_PTHREAD=0)
 
-IF(NOT (Kokkos_ENABLE_CUDA AND WIN32))
+IF((NOT (Kokkos_ENABLE_CUDA AND WIN32)) AND (NOT ("${KOKKOS_CXX_COMPILER_ID}" STREQUAL "Fujitsu")))
```
Comment (dhollman) [inline algorithms/unit_tests/CMakeLists.txt:27]: Elaborate on why this exception is here in a comment, please.

File: Makefile.kokkos [hunk 7]
```diff
@@ -736,6 +737,13 @@ ifeq ($(KOKKOS_INTERNAL_USE_ARCH_ARMV81), 1)
   endif
 endif
 
+ifeq ($(KOKKOS_INTERNAL_USE_ARCH_A64FX), 1)
+  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_A64FX")
+
+  KOKKOS_CXXFLAGS += -march=armv8.2-a+sve
+  KOKKOS_LDFLAGS += -march=armv8.2-a+sve
+endif
+
```
Comment (dhollman) [inline Makefile.kokkos:746]: Do we really need makefile support for _new_ architectures? This is definitely extra maintenance work and it's not really tested in our CI AFAICT.

File: cmake/kokkos_test_cxx_std.cmake [hunk 8]
```diff
@@ -42,13 +42,16 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
   ELSEIF((KOKKOS_CXX_COMPILER_ID STREQUAL "NVIDIA") AND WIN32)
     MESSAGE(STATUS "Using no flag for C++${standard} standard as feature")
     GLOBAL_SET(KOKKOS_CXX_STANDARD_FEATURE "")
+  ELSEIF((KOKKOS_CXX_COMPILER_ID STREQUAL "Fujitsu"))
+    MESSAGE(STATUS "Using no flag for C++${standard} standard as feature")
+    GLOBAL_SET(KOKKOS_CXX_STANDARD_FEATURE "")
```
Comment (dhollman) [inline cmake/kokkos_test_cxx_std.cmake:47]: How is this going to work for C++17 and later? Why isn't there a flag for this?

File: cmake/kokkos_test_cxx_std.cmake [hunk 9]
```diff
@@ -42,13 +42,16 @@ FUNCTION(kokkos_set_cxx_standard_feature standard)
...
   ELSE()
     #nope, we can't do anything here
     MESSAGE(WARNING "C++${standard} is not supported as a compiler feature. We will choose custom flags for now, but this behavior has been deprecated. Please open an issue at https://github.com/kokkos/kokkos/issues reporting that ${KOKKOS_CXX_COMPILER_ID} ${KOKKOS_CXX_COMPILER_VERSION} failed for ${KOKKOS_CXX_STANDARD}, preferably including your CMake command.")
     GLOBAL_SET(KOKKOS_CXX_STANDARD_FEATURE "")
   ENDIF()
 
-  IF(NOT WIN32)
+  IF((NOT WIN32) AND (NOT ("${KOKKOS_CXX_COMPILER_ID}" STREQUAL "Fujitsu")))
```
Comment (dhollman) [inline cmake/kokkos_test_cxx_std.cmake:54]: Why is Fujitsu an exception? Elaborate in a comment please.

File: Makefile.kokkos (hunk 7 above)
Comment (crtrott) [reply Makefile.kokkos:746]: Yes we do.

File: cmake/kokkos_test_cxx_std.cmake (hunk 8 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake:47]: because cmake doesnt' know the fujitsu compiler and hence can't figure out its standard flags and feature set. So we are lying to it here and work around this whole thing. And yes a C++ flag will still be added to the CXX flags during the actual build.

File: cmake/kokkos_test_cxx_std.cmake (hunk 9 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake:54]: I don't know. It just is because cmake is special and this makes it work?

File: cmake/kokkos_test_cxx_std.cmake (hunk 8 above)
Comment (crtrott) [reply cmake/kokkos_test_cxx_std.cmake:47]: Also what do I know about how cmake works, this magic weirdness apparently sidesteps some internal cmake settings/tests/stuff which can't handle situations it doesn't know about like the fujitsu compiler. This makes it work ...

File: core/src/impl/Kokkos_TaskQueue_impl.hpp (hunk 4 above)
Comment (crtrott) [reply core/src/impl/Kokkos_TaskQueue_impl.hpp:384]: this only happens with the fujitsu compiler so is related to adding fujistu compiler support

File: core/unit_test/CMakeLists.txt (hunk 2 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:22]: I was suprised we had issues to begin with. And yes somewhere we needed that with some combination of weird compiler (Intel XL or so) and some cmake version.

File: algorithms/unit_tests/CMakeLists.txt (hunk 6 above)
Comment (crtrott) [reply algorithms/unit_tests/CMakeLists.txt:27]: because cmake is weird and I need to do this to make it work?

Comment (dhollman) [reply algorithms/unit_tests/CMakeLists.txt:27]: This can't be just a CMake problem. @jjwilke or is it?

File: cmake/kokkos_test_cxx_std.cmake (hunk 9 above)
Comment (dhollman) [reply cmake/kokkos_test_cxx_std.cmake:54]: Well if we knew why `WIN32` is an exception, then we might have an idea as to why this is also. I'm not super eager to add to the number of things we don't understand about our CMake system if we can at all help it. We can at least document what the error that this gets around looks like

File: algorithms/unit_tests/CMakeLists.txt (hunk 6 above)
Comment (crtrott) [reply algorithms/unit_tests/CMakeLists.txt:27]: it kinda is, because it combines requesting a c++ standard via compile feature (instead of just setting a cxx flag) with checking whether the compiler actually supports it, which it comes back as negative if it doesn't identify the compiler as a known thing (like nvcc on windows, or fujitsu in gener). I guess I can add that as a comment.

**Review by dhollman — APPROVED** [review]
It's a mess, but all of our build system is a mess apparently. In that context, LGTM.

**Comment by crtrott** [comment]
Opened #3620 which is there to address some of the questions here medium term and streamline the process of side stepping cmake mines for unknown compilers.
